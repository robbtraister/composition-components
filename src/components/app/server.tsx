'use strict'

import React, { useContext } from 'react'

import Tree from '../tree'
import compositionContext from '../../contexts/composition'

const polyfills = {
  assign: '(window.Object&&window.Object.assign)',
  fetch: 'window.fetch',
  includes: 'Array.prototype.includes',
  map: 'window.Map',
  promise: 'window.Promise',
  set: 'window.Set'
}

interface ScriptProps {
  name: string
}

const Script = ({ name }: ScriptProps) => (
  <script type="text/javascript" src={`/dist/${name}.js`} defer />
)

interface ServerAppProps {
  id?: string
  static?: boolean
  'single-page'?: boolean
}

export function App(props: ServerAppProps) {
  const { appName = 'app', cache, output, template, tree } = useContext(
    compositionContext
  )
  const {
    id = 'composition-app',
    static: isStatic = false,
    'single-page': singlePage = false
  } = props

  if (isStatic && singlePage) {
    throw new Error('`static` and `single-page` props are mutually-exclusive')
  }

  const contentCache = Object.keys(cache)
    .filter(entry => !(cache[entry] instanceof Promise))
    .filter(entry => entry.startsWith('{"content":'))
    .reduce((rollup, entry) => {
      rollup[entry] = cache[entry]
      return rollup
    }, {})

  return (
    <>
      {!isStatic && (
        <>
          <script
            type="text/javascript"
            dangerouslySetInnerHTML={{
              __html: `if(!(${Object.values(polyfills).join(
                '&&'
              )}))document.write('<script type="text/javascript" src="/dist/polyfills.js" defer=""><\\/script>');`
            }}
          />
          <Script name="runtime" />
          <Script name="engine" />
          <Script name={singlePage ? `combinations/${output}` : appName} />
        </>
      )}
      <div id={id}>
        <Tree />
      </div>
      {!isStatic && (
        <script
          type="text/javascript"
          dangerouslySetInnerHTML={{
            __html: [
              `window.Composition=window.Composition||{}`,
              `Composition.id=${JSON.stringify(id)}`,
              `Composition.singlePage=${JSON.stringify(singlePage)}`,
              `Composition.cache=${JSON.stringify(contentCache)}`,
              singlePage
                ? `Composition.template=${JSON.stringify(template)}`
                : null,
              singlePage ? `Composition.tree=${JSON.stringify(tree)}` : null
            ]
              .filter(c => c)
              .join(';')
          }}
        />
      )}
    </>
  )
}

export default App
