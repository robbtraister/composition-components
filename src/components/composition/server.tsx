'use strict'

import React from 'react'
import { StaticRouter } from 'react-router'

import { Common } from './common'

export function Composition(props: Composition.ServerCompositionProps) {
  const { children, routerContext = {}, ...contextValue } = props
  return (
    <StaticRouter location={props.location} context={routerContext}>
      <Common value={contextValue}>{children}</Common>
    </StaticRouter>
  )
}

export default Composition
