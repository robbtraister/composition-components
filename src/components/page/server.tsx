'use strict'

import React from 'react'
import { StaticRouter } from 'react-router'

import { CommonPage } from './common'

export function Page(props: Composition.ServerPageProps) {
  const { children, routerContext = {}, ...contextValue } = props
  return (
    <StaticRouter location={props.location} context={routerContext}>
      <CommonPage value={contextValue}>{children}</CommonPage>
    </StaticRouter>
  )
}

export default Page
