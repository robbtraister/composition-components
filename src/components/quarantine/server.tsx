'use strict'

import debugModule from 'debug'
import React from 'react'
import ReactDOM from 'react-dom/server'

import RenderError from './error'

const debug = debugModule('composition:components:quarantine')

export function Quarantine(node: TreeNode) {
  const { id, type, props, children } = node
  try {
    const element = <>{children}</>
    ReactDOM.renderToString(element)
    return element
  } catch (error) {
    debug('caught component error', { type, id, props, error })
    return <RenderError error={error} />
  }
}

export default Quarantine
