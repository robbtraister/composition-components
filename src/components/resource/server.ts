'use strict'

import { useContext } from 'react'

import render from '../render'

import compositionContext from '../../contexts/composition'

export function useResource({ name, encoding = 'utf8' }: ResourceParams) {
  const { cache = {}, getResource } = useContext(compositionContext)

  const key = JSON.stringify({ resource: name })
  if (!(key in cache)) {
    cache[key] = getResource(name, 'buffer')
      .then(data => {
        cache[key] = data
      })
      .catch(() => {
        cache[key] = null
      })
  }

  return cache[key] instanceof Promise
    ? null
    : /^buffer$/i.test(encoding)
    ? cache[key]
    : cache[key] && cache[key].toString(encoding)
}

export function Resource(
  props: RenderableProps<ResourceParams, ResourceStruct>
) {
  const { name, encoding, ...passThroughProps } = props
  const resource = useResource({ name, encoding })

  return render({ ...passThroughProps, resource })
}

export default Resource
