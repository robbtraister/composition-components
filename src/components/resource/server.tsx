'use strict'

import { promises as fsPromises } from 'fs'
import path from 'path'

import React, { useContext } from 'react'

import compositionContext from '../../contexts/composition'

const cachedFiles = {}
async function getCachedFile(filePath) {
  if (!(filePath in cachedFiles)) {
    try {
      cachedFiles[filePath] = await fsPromises.readFile(filePath)
    } catch (_) {}
  }
  return cachedFiles[filePath]
}

export function useResource(name, encoding = 'utf8') {
  const { cache = {}, projectRoot } = useContext(compositionContext)

  const key = JSON.stringify({ resource: name })
  if (!(key in cache)) {
    cache[key] = getCachedFile(path.join(projectRoot, name))
      .then(data => {
        cache[key] = data
      })
      .catch(() => {
        cache[key] = null
      })
  }

  return cache[key] instanceof Promise
    ? null
    : /^buffer$/i.test(encoding)
    ? cache[key]
    : cache[key] && cache[key].toString(encoding)
}

export function Resource(props: {
  name: string
  children?: React.ElementType | React.ElementType[]
  component?: React.ComponentType<{ resource: any }>
  render?: Function
}) {
  const { children, component: Component, render, ...resourceProps } = props
  const { name, ...passThroughProps } = resourceProps
  const resource = useResource(name)

  if (Component) {
    return <Component {...passThroughProps} resource={resource} />
  } else if (render) {
    return render({ ...passThroughProps, resource })
  } else if (children) {
    return []
      .concat(children || [])
      .map((Child, index) => (
        <Child key={index} {...passThroughProps} resource={resource} />
      ))
  }
}

export default Resource
