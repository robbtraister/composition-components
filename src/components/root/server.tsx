'use strict'

import React from 'react'
import { StaticRouter } from 'react-router'

import { CommonRoot } from './common'
import { ServerRootProps } from './types'

export function Root(props: ServerRootProps) {
  const { children, routerContext = {}, ...contextValue } = props
  return (
    <StaticRouter location={props.location} context={routerContext}>
      <CommonRoot value={contextValue}>{children}</CommonRoot>
    </StaticRouter>
  )
}

export default Root
