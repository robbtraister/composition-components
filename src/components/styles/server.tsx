'use strict'

import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

import React, { useContext } from 'react'

import compositionContext from '../../contexts/composition'

const readFile = promisify(fs.readFile.bind(fs))

const StyledComponents = 'styled-components'

interface StylesProps {
  inline?: boolean
}
interface StyleProps {
  name: string
}

const cachedFiles = {}
function getCachedFile(filePath) {
  cachedFiles[filePath] = cachedFiles[filePath] || readFile(filePath)
  return cachedFiles[filePath]
}

export const ServerStyles = ({ inline, ...props }: StylesProps) => {
  const { cache = {}, projectRoot } = useContext(compositionContext)

  const Style = inline
    ? function Style({ name }: StyleProps) {
        const key = JSON.stringify({ styles: name })
        if (key in cache) {
          return cache[key]
        }
        cache[key] = getCachedFile(
          path.join(projectRoot, 'build', 'dist', `${name}.css`)
        )
          .then(data => {
            cache[key] = (
              <style {...props} dangerouslySetInnerHTML={{ __html: data }} />
            )
          })
          .catch(() => {
            cache[key] = null
          })
        return null
      }
    : function Style({ name }: StyleProps) {
        return <link {...props} href={`/dist/${name}.css`} rel="stylesheet" />
      }

  return (
    <>
      <Style name="site" />
      <Style name="app" />
      <StyledComponents />
    </>
  )
}

export default ServerStyles
