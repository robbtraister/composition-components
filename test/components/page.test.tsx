'use strict'

/* global expect, test */

import React from 'react'
import { renderToString } from 'react-dom/server'

import App from '../../src/components/app'
import Composition from '../../src/components/composition'
import Styles from '../../src/components/styles'

import ResourceHandler from '../../src/utils/resources'

import { useComponentContext } from '../../src/contexts/component'

const resourceHandler = new ResourceHandler({ projectRoot: '.' })

const x = 3
function Fail() {
  if (x === 3) {
    throw new Error('fail')
  }
  return <React.Fragment />
}

function Success({ children }: { children: React.ElementType }) {
  const { id, type } = useComponentContext()
  return (
    <div data-type={type} data-id={id}>
      {children}
    </div>
  )
}

function getComponent(type: string): React.ComponentType {
  return type === 'fail' ? Fail : Success
}

const tree = {
  type: 'div',
  id: 'abc',
  children: [
    {
      type: 'fail',
      id: 'def',
      children: [
        {
          type: 'div',
          id: 'ghi'
        }
      ]
    },
    {
      type: 'div',
      id: 'xyz'
    }
  ]
}

test('Page Component', () => {
  const html = renderToString(
    <Composition
      getComponent={getComponent}
      output="default"
      template="test"
      tree={tree}
      quarantine>
      <html>
        <head>
          <title>test</title>
          <Styles />
        </head>
        <body>
          <App single-page />
        </body>
      </html>
    </Composition>
  )
  expect(html).toMatchSnapshot()
})

test('Inline Styles', () => {
  const html = renderToString(
    <Composition
      getComponent={getComponent}
      getResource={resourceHandler.readResourceFile.bind(resourceHandler)}
      output="default"
      template="test"
      tree={tree}
      quarantine>
      <html>
        <head>
          <title>test</title>
          <Styles inline />
        </head>
        <body>
          <App single-page />
        </body>
      </html>
    </Composition>
  )
  expect(html).toMatchSnapshot()
})
