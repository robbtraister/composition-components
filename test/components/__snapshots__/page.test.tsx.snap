// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline Styles 1`] = `"<html><head><title>test</title><style id=\\"composition-site-styles\\"></style><style id=\\"composition-app-styles\\"></style><composition:styled-components></composition:styled-components></head><body><script type=\\"text/javascript\\">if(!((window.Object&&window.Object.assign)&&window.fetch&&Array.prototype.includes&&window.Map&&window.Promise&&window.Set))document.write('<script type=\\"text/javascript\\" src=\\"/dist/polyfills.js\\" defer=\\"\\"><\\\\/script>');</script><script type=\\"text/javascript\\" src=\\"/dist/runtime.js\\" defer=\\"\\"></script><script type=\\"text/javascript\\" src=\\"/dist/engine.js\\" defer=\\"\\"></script><script type=\\"text/javascript\\" src=\\"/dist/combinations/default.js\\" defer=\\"\\"></script><div id=\\"composition-app\\"><div data-type=\\"div\\" data-id=\\"abc\\"><div data-error-component=\\"fail\\" data-error-id=\\"def\\" data-error-message=\\"fail\\"></div><div data-type=\\"div\\" data-id=\\"xyz\\"></div></div></div><script type=\\"text/javascript\\">window.Composition=window.Composition||{};Composition.cache={};Composition.id=\\"composition-app\\";Composition.method=\\"render\\";Composition.singlePage=true;Composition.template=\\"test\\";Composition.tree={\\"type\\":\\"div\\",\\"id\\":\\"abc\\",\\"children\\":[{\\"type\\":\\"fail\\",\\"id\\":\\"def\\",\\"children\\":[{\\"type\\":\\"div\\",\\"id\\":\\"ghi\\"}]},{\\"type\\":\\"div\\",\\"id\\":\\"xyz\\"}]}</script></body></html>"`;

exports[`Page Component 1`] = `"<html><head><title>test</title><link id=\\"composition-site-styles\\" href=\\"/dist/site.css\\" rel=\\"stylesheet\\"/><link id=\\"composition-app-styles\\" href=\\"/dist/app.css\\" rel=\\"stylesheet\\"/><composition:styled-components></composition:styled-components></head><body><script type=\\"text/javascript\\">if(!((window.Object&&window.Object.assign)&&window.fetch&&Array.prototype.includes&&window.Map&&window.Promise&&window.Set))document.write('<script type=\\"text/javascript\\" src=\\"/dist/polyfills.js\\" defer=\\"\\"><\\\\/script>');</script><script type=\\"text/javascript\\" src=\\"/dist/runtime.js\\" defer=\\"\\"></script><script type=\\"text/javascript\\" src=\\"/dist/engine.js\\" defer=\\"\\"></script><script type=\\"text/javascript\\" src=\\"/dist/combinations/default.js\\" defer=\\"\\"></script><div id=\\"composition-app\\"><div data-type=\\"div\\" data-id=\\"abc\\"><div data-error-component=\\"fail\\" data-error-id=\\"def\\" data-error-message=\\"fail\\"></div><div data-type=\\"div\\" data-id=\\"xyz\\"></div></div></div><script type=\\"text/javascript\\">window.Composition=window.Composition||{};Composition.cache={};Composition.id=\\"composition-app\\";Composition.method=\\"render\\";Composition.singlePage=true;Composition.template=\\"test\\";Composition.tree={\\"type\\":\\"div\\",\\"id\\":\\"abc\\",\\"children\\":[{\\"type\\":\\"fail\\",\\"id\\":\\"def\\",\\"children\\":[{\\"type\\":\\"div\\",\\"id\\":\\"ghi\\"}]},{\\"type\\":\\"div\\",\\"id\\":\\"xyz\\"}]}</script></body></html>"`;
